subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ExecuteStep.java

```diff
@@ -16,21 +16,56 @@
 
 package org.gradle.internal.execution.steps;
 
+import com.google.common.collect.ImmutableListMultimap;
 import org.gradle.internal.Try;
 import org.gradle.internal.execution.ExecutionOutcome;
 import org.gradle.internal.execution.IncrementalChangesContext;
 import org.gradle.internal.execution.Result;
 import org.gradle.internal.execution.Step;
+import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.execution.history.changes.ExecutionStateChanges;
+import org.gradle.internal.execution.history.changes.InputChangesInternal;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Optional;
 
 public class ExecuteStep<C extends IncrementalChangesContext> implements Step<C, Result> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExecuteStep.class);
+
     @Override
     public Result execute(C context) {
-        ExecutionOutcome outcome = context.getWork().execute(context.getChanges());
+        UnitOfWork work = context.getWork();
+        InputChangesInternal inputChanges = work.isIncremental()
+            ? determineInputChanges(work, context)
+            : null;
+
+        ExecutionOutcome outcome = work.execute(inputChanges);
         return new Result() {
             @Override
             public Try<ExecutionOutcome> getOutcome() {
                 return Try.successful(outcome);
             }
         };
     }
+
+    private InputChangesInternal determineInputChanges(UnitOfWork work, IncrementalChangesContext context) {
+        Optional<ExecutionStateChanges> changes = context.getChanges();
+        if (!changes.isPresent()) {
+            throw new UnsupportedOperationException("Cannot use input changes when input tracking is disabled.");
+        }
+
+        ImmutableListMultimap<Object, String> incrementalInputs = determineIncrementalInputs(work);
+        InputChangesInternal inputChanges = changes.get().getInputChanges(incrementalInputs);
+        if (!inputChanges.isIncremental()) {
+            LOGGER.info("All input files are considered out-of-date for incremental {}.", work.getDisplayName());
+        }
+        return inputChanges;
```

Comment:
I think we should use Java 8 syntax here. Something like:

```java
return context.getChanges()
    .map(changes -> {
        InputChangesInternal inputChanges = changes.getInputChanges(determineIncrementalInputs(work));
        if (!inputChanges.isIncremental()) {
            LOGGER.info("All input files are considered out-of-date for incremental {}.", work.getDisplayName());
        }
        return inputChanges;
    })
    .orElseThrow(() -> throw new UnsupportedOperationException("Cannot use input changes when input tracking is disabled."));
```

Reply:
Done.